Out-Default incorrectly staying in transcript state

SteveL-MSFT opened the pull request:
Out-Default was incorrectly caching transcription state as the subsequent use was caching the already changed state

Out-Default needed to implement IDispose pattern so transcription state reverts back on exception in pipeline

Addresses https://github.com/PowerShell/PowerShell/issues/3390

<!--

If you are a PowerShell Team member, please make sure you choose the Reviewer(s) and Assignee for your PR.
If you are not from the PowerShell Team, you can leave the fields blank and the Maintainers will choose them for you. If you are familiar with the team, feel free to mention some Reviewers yourself.

For more information about the roles of Reviewer and Assignee, refer to CONTRIBUTING.md.

-->

SteveL-MSFT committed:
Out-Default was incorrectly caching transcription state as the subsequent use was caching the already changed state
Out-Default needed to implement IDispose pattern so transcription state reverts back on exception in pipeline

PetSerAl commented on src/System.Management.Automation/commands/utility/FormatAndOutput/out-console/OutConsole.cs:
```diff
@@ -143,15 +148,30 @@ protected override void EndProcessing()
 
             if (Transcript)
             {
                 Host.UI.TranscribeOnly = _savedTranscribeOnly;
```
You ends up restoring `TranscribeOnly` even if you not capture it before.

SteveL-MSFT replied:
Good catch

PetSerAl commented on src/System.Management.Automation/commands/utility/FormatAndOutput/out-console/OutConsole.cs:
```diff
@@ -79,9 +79,14 @@ protected override void BeginProcessing()
+
             if (Transcript)
             {
+                _transcriptStateChanged = true;
```
You never set `_transcriptStateChanged` back to `false`. So `Out-Default` stop capturing `TranscribeOnly` after first call with `-Transcript` switch.

SteveL-MSFT replied:
`Out-Default` lifetime only exists in that pipeline and typically should be the last thing in the pipeline as nothing is emitted if used with `-Transcript`.  Is there a case that's broken?

PetSerAl replied:
I am not going to say that nested or intersected `Out-Default -Transcript` are useful scenario, but if you are not support it, then (taking in account that `OutDefaultCommand` is the only place where `TranscribeOnly` modified) the only value that `_savedTranscribeOnly` will have is `false`. So, why bother and define variable, which will be always `false`?

PetSerAl reviewed (commented):
It ends up `Out-Default -Transcript` just set `TranscribeOnly = true` on start (`BeginProcessing`) and `TranscribeOnly = false` on end (`Dispose`).

SteveL-MSFT committed:
only restore Host.UI.TranscribeOnly if state had changed

PetSerAl commented on src/System.Management.Automation/commands/utility/FormatAndOutput/out-console/OutConsole.cs:
```diff
@@ -143,15 +148,30 @@ protected override void EndProcessing()
+        private bool _disposed = false;
         private ArrayList _outVarResults = null;
         private bool _savedTranscribeOnly = false;
+        private static bool _transcriptStateChanged = false;
```
You define `_transcriptStateChanged` as `static`, so it affect entire `AppDomain`. That might be not good when multiple PowerShell `Runspace`s share same `AppDomain`.

SteveL-MSFT committed:
added addtional test for normal use case for out-default -transcript
addressed feedback from code review on using static

SteveL-MSFT commented:
@PetSerAl I appreciate the feedback.  I rethought about how BeginProcessing, ProcessRecord, and EndProcessing work and I think I have a solution that remedies your concerns

lzybkr commented on src/System.Management.Automation/commands/utility/FormatAndOutput/out-console/OutConsole.cs:
```diff
@@ -102,6 +98,7 @@ protected override void ProcessRecord()
         {
             if (Transcript)
             {
+                Host.UI.TranscribeOnly = true;
```
Just asking - but should TranscribeOnly be a counter that is incremented each time it's true, and decremented in Dispose?

I'm not sure this is the correct semantic, but it seems right - like it shouldn't be possible to turn it off as long as somebody wants it on.

SteveL-MSFT replied:
I only learned that `Out-Default` had a `-Transcript` switch when this issues was opened and I started looking into it (documentation and tests are also lacking).

Best as I can tell, although you can have nested `Start-Transcript`, `Out-Default` is only used at the end of a pipeline, so the test case of `Out-Default -Transcript | Out-Default -Transcript` I don't think represents a real use case as the intent of `-Transcript` (as far as I can tell) is to suppress output to the pipeline and only send it to the transcript.  However, it got the console stuck in transcript mode, so I thought it was worth fixing (the other issue with exception being raised is a real use case).

So I don't think we need a counter here as it should really just toggle from `true` and `false` when this cmdlet is used.

PetSerAl replied:
> as the intent of -Transcript (as far as I can tell) is to **suppress output to the pipeline**

Sadly it is exact opposite of how `Out-Default -Transcript` is implemented. By default `Out-Default` does not output anything, but display input on host. With `-Transcript` it pass thru the input, but making following `Out-Default`/`Out-Host` to not display their input:
```powershell
1..10 | Out-Default -Transcript | Out-Host
```
But by replacing `Out-Host` with `Write-Host` you can see, that output is really here:
```powershell
1..10 | Out-Default -Transcript | Write-Host
```

SteveL-MSFT replied:
Looking at the code, the only thing `TranscribeOnly` does is [not write to the console] (https://github.com/PowerShell/PowerShell/blob/02b5f357a20e6dee9f8e60e3adb9025be3c94490/src/System.Management.Automation/commands/utility/FormatAndOutput/out-console/ConsoleLineOutput.cs#L324) so I mispoke when I said `pipeline` (should have said `host`).

PetSerAl commented:
@SteveL-MSFT I can not think of any problematic cases now, at least `Out-Default` that PowerShell console host implicitly adds to the pipeline will restore `TranscribeOnly` to `false` at the end of command.

PetSerAl commented on src/System.Management.Automation/commands/utility/FormatAndOutput/out-console/OutConsole.cs:
```diff
@@ -43,7 +43,7 @@ protected override void ProcessRecord()
     /// </summary>
     [Cmdlet(VerbsData.Out, "Default", HelpUri = "https://go.microsoft.com/fwlink/?LinkID=113362", RemotingCapability = RemotingCapability.None)]
-    public class OutDefaultCommand : FrontEndCommandBase
+    public class OutDefaultCommand : FrontEndCommandBase, IDisposable
```
Do you really need explicitly state `IDisposable` here? `FrontEndCommandBase` already implement `IDisposable` and it is not like you need to reimplement `IDisposable.Dispose()` here.

SteveL-MSFT replied:
That was a leftover from before I realized FrontEndCommandBase implements IDisposable.  Will remove.

PetSerAl commented on src/System.Management.Automation/commands/utility/FormatAndOutput/out-console/OutConsole.cs:
```diff
@@ -143,13 +140,23 @@ protected override void EndProcessing()
+        /// <summary>
+        /// Revert transcription state on Dispose
+        /// </summary>
+        protected override void InternalDispose()
```
Are not you supposed to call `base.InternalDispose()` at some point of overriding method?

SteveL-MSFT replied:
You're correct.  Will fix.

SteveL-MSFT committed:
removed unnecessary IDisposable as FrontEndCommandBase already inherits
added missing call to base.InternalDispose()

SteveL-MSFT commented:
@PetSerAl @lzybkr any other feedback preventing signoff?

PetSerAl commented:
@SteveL-MSFT I do not see any obvious wrong things. I still can construct the case where host stuck in `TranscribeOnly` state, but it looks more likely shooting your own leg:
```powershell
begin{
    $sp = {Out-Default -Transcript}.GetSteppablePipeline()
    $sp.Begin($false)
    $sp.Process()
}
end{
    $sp.End()
    $sp.Dispose()
}
```
Possible, the only way to prevent this is to implement `TranscribeOnly` as counter as @lzybkr suggest.
Maybe you should take into account possibility, that cmdlet can be instantiated not for execution, but for parameter/argument completion (not sure if it is the case here as `Out-Default` does not implement dynamic parameters). In that case `BeginProcessing` will not be invoked before disposing, so you will end restoring `TranscribeOnly` when you are not save it.

SteveL-MSFT commented:
I'll rework it to use a counter.  Thanks.

SteveL-MSFT committed:
Changed TranscribeOnly member to a count rather than a bool so nested -Trascript stays in TranscibeOnly
Add try..finally in InternalDispose to ensure TranscribeOnlyCount gets decremented

PetSerAl commented on src/System.Management.Automation/commands/utility/FormatAndOutput/out-console/OutConsole.cs:
```diff
@@ -98,7 +96,7 @@ protected override void ProcessRecord()
             if (Transcript)
             {
-                Host.UI.TranscribeOnly = true;
+                Host.UI.TranscribeOnlyCount++;
```
You are doing this in process block, once for each input item, not once per `Out-Default` invocation.

SteveL-MSFT replied:
Moving back to BeginProcessing()

PetSerAl commented on src/System.Management.Automation/commands/utility/FormatAndOutput/out-console/OutConsole.cs:
```diff
@@ -153,13 +151,15 @@ protected override void InternalDispose()
 
             base.InternalDispose();
-            Host.UI.TranscribeOnly = _savedTranscribeOnly;
+            if (--Host.UI.TranscribeOnlyCount < 0)
```
You are not checking if you actually increment `TranscribeOnlyCount` before. That lead to situation where every call to `Out-Default` reduce counter. Including ones without `-Transcript`. Or where begin block of `Out-Default` does not have opportunity to execute:
```powershell
& {begin{throw}} | Out-Default -Transcript
```

SteveL-MSFT replied:
Will fix

PetSerAl commented on src/System.Management.Automation/commands/utility/FormatAndOutput/out-console/OutConsole.cs:
```diff
@@ -143,15 +143,29 @@ protected override void EndProcessing()
+                return;
+            }
+
+            base.InternalDispose();
```
Maybe wrap this with `try`/`finally`, so you have opportunity to clean up even if `base.InternalDispose()` throw exception.

SteveL-MSFT replied:
Sounds good

PetSerAl commented:
To think about that. Should not it also properly revert `TranscribeOnly` even if `Dispose` is not called, but `Out-Default` collected by garbage collector?
```powershell
$sp = {Out-Default -Transcript}.GetSteppablePipeline()
$sp.Begin($false)
$sp = $null
[GC]::Collect()
```
I have [idea](https://github.com/PowerShell/PowerShell/commit/58a799652ee8718dd3cc5566747f1192cf54d531) about implementation. Can you look at it?

SteveL-MSFT commented:
@PetSerAl I'll look at this today

SteveL-MSFT commented:
@PetSerAl I stepped through the debugger and BeginProcessing is always called and ProcessRecord is called where InputObject is empty.  Finally, Dispose() is called.  In both your samples, once the script is complete, Host.UI.TranscribeOnlyCount = 0.

PetSerAl commented:
@SteveL-MSFT Not sure, what exact code you are referencing as my samples, but I use following code for test:
```powershell
PS> Add-Type @‘
>>>     using System;
>>>     using System.Management.Automation;
>>>     [Cmdlet("Test", "Methods")]
>>>     public class TestMethods : PSCmdlet, IDisposable {
>>>         public TestMethods() {
>>>             Console.WriteLine("Constructor");
>>>         }
>>>         ~TestMethods() {
>>>             Console.WriteLine("Destructor");
>>>         }
>>>         protected override void BeginProcessing() {
>>>             Console.WriteLine("Begin");
>>>         }
>>>         protected override void ProcessRecord() {
>>>             Console.WriteLine("Process");
>>>         }
>>>         protected override void EndProcessing() {
>>>             Console.WriteLine("End");
>>>         }
>>>         protected override void StopProcessing() {
>>>             Console.WriteLine("Stop");
>>>         }
>>>         public void Dispose() {
>>>             Console.WriteLine("Dispose");
>>>         }
>>>     }
>>> ’@ -PassThru | Select-Object -First 1 -ExpandProperty Assembly | Import-Module
PS> $sp = {Test-Methods}.GetSteppablePipeline();
Constructor
PS> $sp.Begin($false)
Begin
PS> $sp = $null
PS> [GC]::Collect()
Destructor
PS>
```
As you can see, `Dispose` is not called here.

SteveL-MSFT commented:
@PetSerAl, by samples, I mean:

```powershell
begin{
    $sp = {Out-Default -Transcript}.GetSteppablePipeline()
    $sp.Begin($false)
    $sp.Process()
}
end{
    $sp.End()
    $sp.Dispose()
}
```

and 

```powershell
$sp = {Out-Default -Transcript}.GetSteppablePipeline()
$sp.Begin($false)
$sp = $null
[GC]::Collect()
```

In both of these cases, the `Out-Default` cmdlet's Begin/Process/Dispose are called (where `-Transcript` is passed).

PetSerAl commented:
@SteveL-MSFT The first sample got fixed, when you implement `TranscribeOnly` as a counter. But second one is still lock console in transcribe only mode, when I test in on [this](https://github.com/PowerShell/PowerShell/pull/3392/commits/a2069d6f9c27252d73f81c7d89735225fee491dd) commit:
```powershell
PS> $sp = {Out-Default -Transcript}.GetSteppablePipeline()
PS> $sp.Begin($false)
PS> $sp = $null
PS> [GC]::Collect()
PS> 1..10
PS> [System.Management.Automation.Host.PSHostUserInterface].GetProperty('TranscribeOnlyCount', [System.Reflection.BindingFlags]'Instance, NonPublic').GetValue($Host.UI) | Write-Host
1
PS>
```

SteveL-MSFT committed:
added test case where garbage collection happens

SteveL-MSFT commented:
@PetSerAl your proposed change looks good to me, can you submit that as a PR and include the tests I added to this PR?  I'll close this one

PetSerAl commented:
@SteveL-MSFT OK. I can do that.

SteveL-MSFT commented:
@PetSerAl thanks for your help on this
